concepts/functions/introduction.md

```diff
@@ -1,46 +1,113 @@
+// One parameter
+func PrintHelloName(name string) {
+  fmt.Println("Hello " + name)
+}
+```
 
-Go is a statically-typed language, which means that everything has a type at compile-time. Assigning a value to a name is referred to as defining a variable. A variable can be defined either by explicitly specifying its type, or by assigning a value to have the Go compiler infer its type based on the assigned value.
+## Return Values
+
+The function arguments are followed by zero or more return values which must also be explicitly typed.
+Single return values are left bare, multiple return values are wrapped in parenthesis.
+Values are returned to the calling code from functions using the `return` keyword.
+There can be multiple `return` statements in a function.
+The execution of the function ends as soon as it hits one of those `return`s.
+If multiple values are to be returned from a function, they are comma seperated.
 
 ```go
-var explicit int // Explicitly typed
-implicit := 10   // Implicitly typed
+func Hello(name string) string {
+  return "Hello " + name
+}
+
+func HelloAndGoodbye(name string) (string. string) {
+  return "Hello " + name, "Goodbye " + name
+}
 ```
 
-The value of a variable can be assigned using the `:=` and updated using the `=`. Once defined, a variable's type can never change.
+## Invoking Functions
+
+Invoking a function is done by specifying the function name and passing arguments for each of the function's parameters in parenthesis.
 
 ```go
-count := 1 // Assign initial value
-count = 2  // Update to new value
+import "fmt"
 
-// Compiler error when assigning different type
-// count = false
+// No parameters, no return value
+func PrintHello() {
+    fmt.Println("Hello")
+}
+// Called like this:
+PrintHello()
+
+// One parameter, one return value
+func Hello(name string) string {
+  return "Hello " + name
+}
+// Called like this:
+greeting := Hello("Dave")
+
+// Multiple parameters, multiple return values
+func SumAndMultiply(a, b int) (int, int) {
+    return a+b, a*b
+}
+// Called like this:
+aplusb, atimesb := SumAndMultiply(a, b)
 ```
 
-A function can have zero or more parameters. All parameters must be explicitly typed, there is no type inference for parameters. A function can also have multiple return values which must also be explicitly typed. Values are returned from functions using the `return` keyword. To allow a function to be called by code in other packages, the name of the function must start with a capital letter.
+## Named Return Values and Naked Return
 
-```go
-package greeting
+As well as parameters, return values can optionally be named.
+If named return values are used, a `return` statement without arguments will return those values, this is known as a 'naked' return.
 
-// Hello is a public function
-func Hello(name string) string {
-    return hello(name)
+```go
+func SumAndMultiplyThenMinus(a, b, c int) (sum, mult int) {
+    sum, mult = a+b, a*b
+    sum -= c
+    mult -= c
+    return
 }
```

Comment:
This example uses multiple returns, which is something that is introduced in the `card-tricks` exercise. The dependencies here are tricky.

Reply:
The other option is to return a `struct` type - the `structs` concept only depends on `numbers` and `conditonals-if` so this doesn't seem unreasonable...

Reply:
Per @junedev we're OK with multiple returns here, and will eventually remove them from `card-tricks` - resolving this.